### PR TITLE
Process all incoming data, also in case of bursty sends

### PR DESCRIPTION
**Pull request by VeijoPesonen**
The driver must continue processing incoming data as long as there is
data available AND stop only when expected data is received. Earlier
driver could stop receiving even when data was available.

When sending data in bursts - multiple sends before receiving any -
there is need to drain RX buffer to avoid overflow in the middle. This
is a polling driver after all.

These changes wont make the receiver to work perfectly but will make the
situation better.

### CODE REVIEW

**Comment by kjbracey-arm** [comment]
Getting increasingly nervous about this. I'd feel happier of process_oob wasn't broken so it blocked for over a second. Going deeper down this path without fixing process_oob is increasing the chance that you're "improving" performance at the expense of blocking the CPU when you're not supposed to be.

Have you instrumented that at all - the time you're spending in these calls? Maybe you're not blocking for a second, but code inspection suggests you probably are.

Not quite sure what recv's up to now is now - "process OOBs until empty, check for packets, if no packets try to process one more OOB, just in case something's arrived in the last 5 microseconds"? Does that nearly instantaneous recheck really affect anything?

**Comment by VeijoPesonen** [comment]
Yeah, my mistake. Fixed based on @kjbracey-arm's comments

File: ESP8266/ESP8266.cpp [hunk 1]
```diff
@@ -286,7 +286,7 @@ bool ESP8266::send(int id, const void *data, uint32_t amount)
         if (_parser.send("AT+CIPSEND=%d,%lu", id, amount)
             && _parser.recv(">")
             && _parser.write((char*)data, (int)amount) >= 0) {
-            _parser.process_oob(); // multiple sends in a row require this
+            while (_parser.process_oob()); // multiple sends in a row require this
```
Comment (SeppoTakalo) [inline ESP8266/ESP8266.cpp]: How long can this block?

Comment (VeijoPesonen) [reply ESP8266/ESP8266.cpp]: It might block for half a second

Comment (VeijoPesonen) [reply ESP8266/ESP8266.cpp]: Same ~0.5 second delay occurs also with ethernet every now and then

**Comment by SeppoTakalo** [comment]
So it is a suboptimal then, but at least somehow stable.

**Comment by kjbracey-arm** [comment]
It just becomes a non-fix when/if process_oob is fixed to not block.

(Although maybe we don't dare touch that, because code like this relies on the block, and put in the non-blocking OOB check separately as recv(NULL)).

**Pull request merged**